### PR TITLE
feat: Story map collaboration

### LINKED ISSUE #1037: Engineering task -Share draft story map with individuals to collaborate

### LINKED ISSUE #1161: Story Maps: clean up invite dialog
## Description
* The placeholder text "Enter one or more email address" is ungrammatical.
suggestion: "Enter email addresses"

* Add one or more people to edit your story map --> Invite people to edit your story map

* Drop "They will receive an email invitation to join this story map."

* Remove "Editors" label from text box.

* Add more space between email address and role

image

### LINKED ISSUE #1163: Story Maps: grammar/typography errors in removal dialog
## Description
In the story map remove user confirmation dialog: 

1. "Remove XYZ from this story map" should have a question mark at the end.
2. `XYZ will no longer be able to edit the “ABC map”.` should be `XYZ will no longer be able to edit the story map “ABC map”.`

### LINKED ISSUE #1164: Story maps: use different text for removing pending users
## Description
Removing a pending user shows this text:
>Remove XYZ from this story map
>XYZ will no longer be able to edit the “Poopy map”.

This isn't accurate, as the user hasn't accepted the invite yet.

Suggestion:
Withdraw XYZ's invitation to this story map?
XYZ will not be able to edit “Poopy map”.

### PR DESCRIPTION
**Pull request by josebui**
## Description

- Added Share dialog
- Added Story map collaboration pending state
- Added support for invites for not registered users

### Checklist
- [x] Corresponding issue has been opened
- [x] New tests added

### Related Issues
Fixes #1037
Fixes https://github.com/techmatters/terraso-product/issues/336
Fixes https://github.com/techmatters/terraso-backend/issues/794
Fixes #1164
Fixes #1163
Fixes #1161

### CODE REVIEW

File: src/collaboration/components/MemberName.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
```
Comment (paulschreiber) [inline src/collaboration/components/MemberName.js:1]: issue: add copyright notice.

File: src/collaboration/collaborationConstants.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+export const MEMBERSHIP_STATUS_APPROVED = 'APPROVED';
```
Comment (paulschreiber) [inline src/collaboration/collaborationConstants.js:1]: issue: add copyright notice.

File: src/collaboration/components/MembershipsList.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
+import React, { useMemo } from 'react';
```
Comment (paulschreiber) [inline src/collaboration/components/MembershipsList.js:1]: issue: add copyright notice.

File: src/tests/uiUtils.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { fireEvent, within } from 'tests/utils';
```
Comment (paulschreiber) [inline src/tests/uiUtils.js:1]: issue: add copyright notice.

File: src/tests/apiUtils.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import * as terrasoApi from 'terraso-client-shared/terrasoApi/api';
```
Comment (paulschreiber) [inline src/tests/apiUtils.js:1]: issue: add copyright notice.

File: src/localization/i18n.js [hunk 6]
```diff
@@ -61,4 +61,11 @@ i18n.services.formatter.add('lowercase', (value, lng, options) => {
   return i18n.t(value).toLowerCase();
 });
 
+i18n.services.formatter.add('andList', (value, lng, options) => {
+  if (value.length === 1) {
+    return value[0];
+  }
+  return `${value.slice(0, -1).join(', ')} ${i18n.t('and')} ${value.slice(-1)}`;
+});
+
```
Comment (paulschreiber) [inline src/localization/i18n.js:70]: suggestion: use `Intl.ListFormat` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat

File: src/localization/locales/en-US.json [hunk 7]
```diff
@@ -981,13 +980,43 @@
...
+        "share_dialog_description": "Add one or more people to edit your story map. <1>Learn more about collaborating on story maps</1>.",
+        "share_dialog_help_url": "https://terraso.org/help/how-to-manage-editors-for-your-story-map/",
+        "share_dialog_confirm_label": "Invite",
+        "share_dialog_cancel_label": "Cancel",
+        "remove_membership": "Remove",
+        "role_collaborator": "Editor",
+        "role_owner": "Owner",
+        "share_dialog_collaborators_title": "People with access",
+        "share_dialog_autocomplete_label": "Editors",
+        "share_dialog_autocomplete_helper_text": "The person you invite will receive an email invitation to join this story map in a few minutes.",
```
Comment (paulschreiber) [inline src/localization/locales/en-US.json:994]:
```suggestion
        "share_dialog_autocomplete_helper_text": "They will receive an email invitation to join this story map.",
```

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: src/localization/i18n.js [hunk 8]
```diff
@@ -61,4 +61,12 @@ i18n.services.formatter.add('lowercase', (value, lng, options) => {
   return i18n.t(value).toLowerCase();
 });
 
+i18n.services.formatter.add('andList', (value, lng, options) => {
```
Comment (paulschreiber) [inline src/localization/i18n.js:64]: question: can we use [i18n's built-in list support?](https://www.i18next.com/translation-function/formatting#list) 
```
// JSON
{
  "intlList": "A list of {{val, list}}"
}

i18next.t('intlList', { val: ['locize', 'i18next', 'awesomeness'] });
// --> A list of locize, i18next, and awesomeness
```

**Review by paulschreiber — CHANGES_REQUESTED** [review]

**Review by paulschreiber — APPROVED** [review]

File: src/storyMap/storyMapService.js [hunk 9]
```diff
@@ -130,3 +145,123 @@ export const deleteStoryMap = ({ storyMap }) => {
...
+    }
+  `);
+
+  return terrasoApi
+    .requestGraphQL(query, {
+      storyMapId: storyMap.storyMapId,
+      storyMapSlug: storyMap.slug,
+      userEmails: emails,
+      userRole,
+    })
```
Comment (shrouxm) [inline src/storyMap/storyMapService.js:179]: nit: in other mutations, we do this which has less code dup (applies to several mutations in this file):
```suggestion
  const query = graphql(`
    mutation addMemberships($input: StoryMapMembershipSaveMutationInput!) {
      saveStoryMapMembership(input: $input) {
        memberships {
          ...collaborationMembershipFields
        }
        errors
      }
    }
  `);

  return terrasoApi
    .requestGraphQL(query, {
      input: {
        storyMapId: storyMap.storyMapId,
        storyMapSlug: storyMap.slug,
        userEmails: emails,
        userRole,
      },
    })
```

**Pull request merged**